prow/bugzilla/client.go

```diff
@@ -220,31 +259,105 @@ func (c *client) GetRootForClone(bug *Bug) (*Bug, error) {
 // GetAllClones returns all the clones of the bug including itself
 // Differs from GetClones as GetClones only gets the child clones which are one level lower
 func (c *client) GetAllClones(bug *Bug) ([]*Bug, error) {
-	return getAllClones(c, bug)
+	bugCache := newBugDetailsCache()
+	return getAllClones(c, bug, bugCache)
 }
-func getAllClones(c Client, bug *Bug) ([]*Bug, error) {
-	root, err := getRootForClone(c, bug)
+
+func getAllClones(c Client, bug *Bug, bugCache *bugDetailsCache) (clones []*Bug, err error) {
+
+	clones = []*Bug{}
+	err = getAllLinkedBugs(c, bug.ID, bugCache, nil)
 	if err != nil {
 		return nil, err
 	}
-	clones, err := getRecursiveClones(c, root)
-	if err != nil {
-		return nil, err
+	cachedBugs := bugCache.list()
+	for index, node := range cachedBugs {
+		if node.Summary == bug.Summary {
+			clones = append(clones, &cachedBugs[index])
+		}
 	}
-	// Getting rid of the bug whose clones we are searching for
-	// we could optimize this (maybe?) if the list turns out to be too long
-	indexOfOriginal := -1
-	for index, clone := range clones {
-		if clone.ID == bug.ID {
-			// Check if there is a more elegant way to do this
-			indexOfOriginal = index
+	return clones, nil
+}
+
+// Parallel implementation for getAllClones - spawns threads to go up and down the tree
+// Also parallelizes the getBug calls if bug has multiple bugs in DependsOn/Blocks
+func getAllLinkedBugs(c Client, bugID int, bugCache *bugDetailsCache, errGroup *errgroup.Group) error {
+	var shouldWait bool
+	if errGroup == nil {
+		shouldWait = true
+		errGroup = new(errgroup.Group)
+	}
+	var err error
+	var bug *Bug
+	bugObj, cacheHasBug := bugCache.get(bugID)
+	if !cacheHasBug {
+		bug, err = c.GetBug(bugID)
+		if err != nil {
+			return err
 		}
+	} else {
+		bug = &bugObj
 	}
-	if indexOfOriginal == -1 {
-		return nil, fmt.Errorf("Original bug not found in list of clones. Error in logic for getRecursiveClones/getRootForClone")
+	errGroup.Go(func() error {
+		return traverseUp(c, bug, bugCache, errGroup)
+	})
+	errGroup.Go(func() error {
+		return traverseDown(c, bug, bugCache, errGroup)
+	})
+
+	if shouldWait {
+		return errGroup.Wait()
 	}
-	clones = append(clones[:indexOfOriginal], clones[indexOfOriginal+1:]...)
-	return clones, nil
+	return nil
+}
+
+func traverseUp(c Client, bug *Bug, bugCache *bugDetailsCache, errGroup *errgroup.Group) error {
+	for _, dependsOnID := range bug.DependsOn {
+		dependsOnID := dependsOnID
+		errGroup.Go(func() error {
+			parentObj, alreadyFetched := bugCache.get(dependsOnID)
+			if alreadyFetched {
+				return nil
+			}
+			parent := &parentObj
+			var err error
+			parent, err = c.GetBug(dependsOnID)
```

Comment:
I believe you can simply do a
```
parentObj, err := c.GetBug(dependsOnID)`
```

Reply:
parentObj is also of type Bug and parent is of type *Bug